### ISSUE TITLE
Fix wrong implementation of compareTo in LocalizedMessages

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#4454]
Based on https://github.com/checkstyle/checkstyle/issues/4409#issuecomment-308021518 and https://github.com/checkstyle/checkstyle/issues/4409#issuecomment-308231553, LocalizedMessages implements Comparable and overrides compareTo method. In the current implementation 

```java
localizedMessage.compareTo(localizedMessage) == 0) == (localizedMessage.equals(localizedMessage)
```

will return false as 'compareTo' method does not take into account all fields which are used in 'equals'.

 'compareTo' method should be fixed to follow the contract with equals.

**Comment by Pimkenini** [comment]
In order to implement 'compareTo' with the same fields as used in 'equals', the fields themselves need to be comparable, but some of the fields used by equals (e.g. 'severityLevel') currently are not. 

Is an arbitrary ordering (by String comparison of the severity level names perhaps) in cases where a tie is broken by 'severityLevel' acceptable, or should comparison of SeverityLevel be implemented as well (wherein 'IGNORE' < 'INFO' < 'WARNING' < 'ERROR')?

**Comment by soon** [comment]
I'd like to take this issue (if it is still unassigned), it is required to implement https://github.com/checkstyle/checkstyle/issues/4409

**Comment by rnveach** [comment]
@romani Has to approve it first before anyone can work on it.

**Comment by soon** [comment]
Some notes about this issue: in the current implementation, the checkstyle skips duplicate messages from the different checks, i.e.:

> [WARN] /home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.java:764: Javadoc comment at column 29 has parse error. Details: no viable alternative at input 'nano' while parsing JAVADOC_INLINE_TAG [JavadocTagContinuationIndentation]
> [WARN] /home/soon/Src/Java/checkstyle-contribution/checkstyle-tester/repositories/guava/guava-tests/test/com/google/common/util/concurrent/JSR166TestCase.java:764: Javadoc comment at column 29 has parse error. Details: no viable alternative at input 'nano' while parsing JAVADOC_INLINE_TAG [NonEmptyAtclauseDescription]

This is because `LocalizedMessages#compareTo` does not take into account the module name. While solving this issue, should we also skip duplicates?

**Comment by soon** [comment]
@romani Could you please approve this issue? I'm comparing the ST output (from the master branch) and MT output (from the #4409 branch) to verify the correctness of the MT changes. However, because of this issue the output of MT implementation is either unstable (the module name for a duplicate messages may be different across the multiple runs) or contains duplicates (because of fixed `compareTo` method which takes into account all fields).

**Comment by rnveach** [comment]
> This is because LocalizedMessages#compareTo does not take into account the module name. While solving this issue, should we also skip duplicates?

These aren't duplicates. Different modules are reporting them.
This issue is at https://github.com/checkstyle/checkstyle/issues/3340 and how we will handle it.

**Comment by arturict** [comment]
I am on it.

**Comment by arturict** [comment]
I looked into this. Making compareTo consistent with equals is mechanically small: keep line, column, module id, source class, and rendered message as the leading criteria, then add the remaining equals fields (column char index, token type, severity, key, bundle, custom message, args) as tie breakers. Ordering of previously distinct violations stays the same.

Running the full suite on that patch shows the real cost. 31 tests fail, and every failure has the same shape: Checker and TreeWalker collect violations in a TreeSet, so pairs that compared as 0 without being equal were silently dropped. With the contract fixed they survive, and for the user they are duplicate report lines: same module, same line and column, same message text, differing only in token type or column char index.

Concrete cases from the suite on current master plus the patch:

- SingleSpaceSeparatorCheckTest.testSpaceErrors expects 31 violations and gets 41. The check logs the same message at the same position once per affected child node.
- LeftCurlyCheckTest.testDefaultLambda expects [17, 24, 27] and gets [17, 17, 24, 27]. The same brace is reported from two token contexts.
- SuppressionXpathFilterExamplesTest and the two Examples tests fail the same way.

So the compareTo fix alone changes user facing output: previously collapsed duplicates get printed several times. Before sending a PR I want to confirm the direction:

1. Fix compareTo and also fix the checks that log the same message twice at the same position (SingleSpaceSeparator, LeftCurly, possibly more). The duplicate logging looks like the actual defect, and report output stays clean.
2. Fix compareTo and de-duplicate at reporting time by the user visible fields (position, severity, module, text), which keeps today's output.
3. Fold this into the wider duplicate handling discussed in #3340.

I prefer option 1 and would start with compareTo plus the SingleSpaceSeparator and LeftCurly logging fixes, each as its own bounded change.

**Comment by romani** [comment]
Good  with 1

**Comment by arturict** [comment]
@romani fyi, I am back at work and school since mid August and have a lot less time than before. I should have said something earlier. I will still look in when I can and pick up smaller things, but a load like before is realistically only possible again around October.

The compareTo fix for this issue is ready locally, I will send the PR when I find a free evening. If anyone wants it earlier, feel free.

**Comment by romani** [comment]
Fix is merged
